### DISCUSSION TITLE
AsDiscrete Transform

### DISCUSSION THREAD

**Discussion by mfaizan-ai** [Project-MONAI/MONAI#3741] [Q&A]
I am facing an issue, I am trying to solve, but I couldn't. 
Here is code:
```py
post_pred =  Compose([EnsureType(), Activations(softmax=True)])
post_label = Compose([EnsureType(), AsDiscrete(to_onehot=2)])

acc_value = torch.eq(y_pred.argmax(dim=1), y)
acc_metric = acc_value.sum().item() / len(acc_value)
y_onehot = [post_label(i) for i in decollate_batch(y)]
 y_pred_act = [post_pred(i) for i in decollate_batch(y_pred)]
auc_metric(y_pred_act, y_onehot)
auc_result = auc_metric.aggregate()
auc_metric.reset()
del y_pred_act, y_onehot
```

The same code is provided [here](https://github.com/Project-MONAI/tutorials/blob/master/3d_classification/torch/densenet_training_dict.py). I have no idea whats going on?

I am facing this error.
AssertionError: One of self.n_classes or n_classes must be an integer
The above exception was the direct cause of the following exception:
raise RuntimeError(f"applying transform {transform}") from e

applying transform <monai.transforms.post.array.AsDiscrete object at 0x7f48e0027a90>

**Comment by wyli** [comment]
the current codebase probably wouldn't generate this error message `AssertionError: One of self.n_classes or n_classes must be an integer`. could you please post the exact error message with the relevant traceback? @faizan1234567

**Reply by mfaizan-ai** [reply]
Environment info
================================
Printing MONAI config...
================================
MONAI version: 0.6.0
Numpy version: 1.19.5
Pytorch version: 1.10.0+cu111
MONAI flags: HAS_EXT = False, USE_COMPILED = False
MONAI rev id: 0ad9e73639e30f4f1af5a1f4a45da9cb09930179

Optional dependencies:
Pytorch Ignite version: 0.4.5
Nibabel version: 3.0.2
scikit-image version: 0.18.3
Pillow version: 7.1.2
Tensorboard version: 2.7.0
gdown version: 4.2.0
TorchVision version: 0.11.1+cu111
ITK version: NOT INSTALLED or UNKNOWN VERSION.
tqdm version: 4.62.3
lmdb version: 0.99
psutil version: 5.4.8
pandas version: 1.1.5
einops version: NOT INSTALLED or UNKNOWN VERSION.

For details about installing the optional dependencies, please visit:
    https://docs.monai.io/en/latest/installation.html#installing-the-recommended-dependencies


================================
Printing system config...
================================
System: Linux
Linux version: Ubuntu 18.04.5 LTS
Platform: Linux-5.4.144+-x86_64-with-Ubuntu-18.04-bionic
Processor: x86_64
Machine: x86_64
Python version: 3.7.12
Process name: python3
Command: ['/usr/bin/python3', '-m', 'ipykernel_launcher', '-f', '/root/.local/share/jupyter/runtime/kernel-52ad6c3a-2889-49f4-8232-463b73021ab2.json']
Open files: []
Num physical CPUs: 1
Num logical CPUs: 2
Num usable CPUs: 2
CPU usage (%): [20.4, 21.9]
CPU freq. (MHz): UNKNOWN for given OS
Load avg. in last 1, 5, 15 mins (%): UNKNOWN for given OS
Disk usage (%): 53.9
Avg. sensor temp. (Celsius): UNKNOWN for given OS
Total physical memory (GB): 12.7
Available memory (GB): 11.3
Used memory (GB): 1.1

================================
Printing GPU config...
================================
Num GPUs: 1
Has CUDA: True
CUDA version: 11.1
cuDNN enabled: True
cuDNN version: 8005
Current device: 0
Library compiled for CUDA architectures: ['sm_37', 'sm_50', 'sm_60', 'sm_70', 'sm_75', 'sm_80', 'sm_86']
GPU 0 Name: Tesla T4
GPU 0 Is integrated: False
GPU 0 Is multi GPU board: False
GPU 0 Multi processor count: 40
GPU 0 Total memory (GB): 14.8
GPU 0 CUDA capability (maj.min): 7.5

**Reply by wyli** [reply]
so you are using `MONAI version: 0.6.0`, the tutorial you linked is for monai 0.8.0. for the v0.6 tutorials, they are at this tag https://github.com/Project-MONAI/tutorials/tree/0.6.0

**Comment by mfaizan-ai** [comment]
ok, I will try it with 0.8.0, then I will let you know.

**Comment by mfaizan-ai** [comment]
![error](https://user-images.githubusercontent.com/61932757/151668752-1aef98e7-3701-47ff-b8d7-776567a3861b.png)

**Reply by wyli** [reply]
this is from MONAI 0.6.0, you can use 
```py
    post_pred = Compose([EnsureType(), Activations(softmax=True)])
    post_label = Compose([EnsureType(), AsDiscrete(to_onehot=True, n_classes=2)])
```

**Comment by mfaizan-ai** [comment]
Thank you so much. The issue has been resolved. 
A question, I think it's irrelevant lol. The models MONAI provides are pretrained? Like densenet121, I was looking for pretrained model on a medical dataset. However, I have not found anything solid. I think pretrained option is available for 2d classification.

**Reply by wyli** [reply]
great!

you are right there isn't currently 3D pretrained, there are some possible ideas discussed here a while ago https://github.com/Project-MONAI/MONAI/issues/2289.
